analyzers/src/SonarAnalyzer.CSharp/Rules/ClassAndMethodName.cs

```diff
@@ -70,10 +71,60 @@ public sealed class ClassAndMethodName : SonarDiagnosticAnalyzer
                 KnownType.System_Runtime_InteropServices_ComImportAttribute,
                 KnownType.System_Runtime_InteropServices_InterfaceTypeAttribute);
 
+        internal static IEnumerable<string> SplitToParts(string name)
+        {
+            var currentWord = new StringBuilder();
+            foreach (var c in name)
+            {
+                if (char.IsUpper(c))
+                {
+                    if (currentWord.Length > 0 && !char.IsUpper(currentWord[currentWord.Length - 1]))
+                    {
+                        yield return currentWord.ToString();
+                        currentWord.Clear();
+                    }
+
+                    currentWord.Append(c);
+                }
+                else if (char.IsLower(c))
+                {
+                    if (currentWord.Length > 1 && char.IsUpper(currentWord[currentWord.Length - 1]))
+                    {
+                        var lastChar = currentWord[currentWord.Length - 1];
+                        currentWord.Length--;
+                        yield return currentWord.ToString();
+                        currentWord.Clear();
+                        currentWord.Append(lastChar);
+                    }
+
+                    currentWord.Append(c);
+                }
+                else
+                {
+                    if (currentWord.Length > 0)
+                    {
+                        yield return currentWord.ToString();
+                        currentWord.Clear();
+                    }
+
+                    yield return c.ToString();
+                }
+            }
+
+            if (currentWord.Length > 0)
+            {
+                yield return currentWord.ToString();
+            }
+        }
+
         protected override void Initialize(SonarAnalysisContext context)
         {
             context.RegisterSyntaxNodeAction(c =>
                 {
+                    if (c.ContainingSymbol.Kind != SymbolKind.NamedType)
```

Comment:
Let`s use RegisterSyntaxNodeAction instead of RegisterSyntaxNodeAction so we can remove the second if with ShouldAnalyze.